mltensor-explainer.md

```diff
@@ -0,0 +1,421 @@
+
+### Support taking a view over an `MLTensor`
+
+This proposal does not include the ability to arbitrarily reinterpret the attributes or contents of an `MLTensor` - in line with the stance that an `MLTensor` represents a tensor rather than an opaque bag of bytes. Tensor operations such as taking a view over a tensor, reinterpreting a tensor as a new shape, or casting its contents to a new data type map to WebNN's `slice`, `reshape`, and `cast` operators, respectively. To reinterpret an `MLTensor`, these tensor operators may be moved _into the graph itself_ or performed in a separate `MLGraph`.
+
+You may want to write the following code, but taking a view over an `MLTensor` is not supported:
+
+```js
+// If creating a slice of an MLTensor in script was supported
+
+const mlTensor = mlContext.createTensor({ dataType: 'int32', shape: [4, 3]});
+
+// Create an input which is half the size of `mlTensor`.
+const operandOfDesiredShape = builder.input('a', { dataType: 'int32', shape: [2, 3] });
+// ... build the rest of the graph using `operandOfDesiredShape`...
+
+// Pass a view over the top half of the tensor as a graph input.
+// NOTE: THIS SUBSCRIPT METHOD DOES THIS EXIST
+const mlTensorView = mlTensor[:2];
+mlContext.dispatch(graph, {'a': mlTensorView}, outputs);
+```
+
+One way to work around this is by inserting a `slice` operation _within the graph_.
+
+```js
+// Workaround which inserts a slice operation within the graph itself
+
+const mlTensor = mlContext.createTensor({ dataType: 'int32', shape: [4, 3]});
+
+// Create an input which is exactly the size of `mlTensor`, then add a slice
+// operation within the graph itself to get the desired view over the input.
+const input = builder.input('a', { dataType: 'int32', shape: [4, 3] });
+const operandOfDesiredShape = builder.slice(input, /*starts=*/[0, 0], /*sizes=*/[2, 3])
+// ... build the rest of the graph using `operandOfDesiredShape`...
+
+// Pass the MLTensor directly.
+mlContext.dispatch(graph, {'a': mlTensor}, outputs);
+```
+
+It's possible that demand may emerge for some mechanism to copy data between `MLTensor`s, though that is currently not planned. This may be worked around by creating another graph with an `identity` operation.
+
+### Support passing an `MLTensor` as a `constant()`
+
+There may be an appetite for a mechanism to stream constant weights to the ML context without needing to go through an `ArrayBuffer`, but whether that mechanism should use the `MLTensor` interface is not yet clear. It's not a natural fit because an `MLTensor` is scoped to an `MLContext`, whereas a constant `MLOperand` is scoped to a single `MLGraphBuilder` (and subsequently the `MLGraph` it builds), since sharing constant data across graphs is [not reliably supported on some platforms](https://github.com/webmachinelearning/webnn/issues/614#issuecomment-2021581363).
```

Comment:
Can we actually share constant data between `MLGraph`s, though? The linked comment suggests not, at least on DML

If the weights cannot be shared across graphs, what's the advantage of using `MLTensor` as opposed to something like https://crrev.com/c/5689303?

Reply:
DML allows CPU-visible buffers containing the original constant data to be re-used between graphs - the restriction is _only_ on the GPU-side destination buffer because it is owned by DML. 

Technically, we could disallow "owned by DML" and re-use the same `MLTensor` w/o any copy. 

So I think we should consider `constant MLBuffer`, at-least to avoid the CPU copies.

Reply:
That would essentially treat all constants as inputs, no?

Regardless, I propose we continue this discussion on another issue (would you like to file one?). This paragraph says that:

> [`MLTensor` is] not a natural fit because an `MLTensor` is scoped to an `MLContext`, whereas a constant `MLOperand` is scoped to a single `MLGraphBuilder` 

If we decide that constants may be scoped to an `MLContext`, then this assumption no longer holds and we can certainly revisit this stance. But for now constants are scoped to an `MLGraph[Builder]` so `MLTensor` is not currently appropriate. Thoughts?

Reply:
Resolving, no action needed.